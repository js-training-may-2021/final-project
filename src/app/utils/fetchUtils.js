import axios from 'axios';
import _ from 'lodash';

const isDev = process.env.NODE_ENV === 'development';
const pathApi = isDev ? 'http://localhost:3000' : '/api/v1';
export const limit = 20;

const fetchPokemons = (page, query, attr, order) => axios.get(`${pathApi}/pokemons?q=${query}&_page=${page}&_limit=${limit}&_sort=${attr}&_order=${order}`);
const normalizedQuery = (query) => _.trimStart(query, '#');

export const mappingFetchPokemons = {
  idAsc: (page, query) => fetchPokemons(page, normalizedQuery(query), 'id', 'asc'),
  idDesc: (page, query) => fetchPokemons(page, normalizedQuery(query), 'id', 'desc'),
  nameAsc: (page, query) => fetchPokemons(page, normalizedQuery(query), 'name', 'asc'),
  nameDesc: (page, query) => fetchPokemons(page, normalizedQuery(query), 'name', 'desc'),
};

export const fetchPokemonById = (id) => axios.get(`${pathApi}/pokemons?id=${id}`);

export const fetchData = () => (
  axios({
    method: 'post',
    url: '/api/v1/data',
    data: { token: localStorage.token },
    timeout: 4000,
  })
);

export const fetchCatch = (pokemon) => axios({
  method: 'post',
  url: '/api/v1/catchPokemon',
  data: { ...pokemon, token: localStorage.token },
  timeout: 4000,
});

export const fetchLogin = (data) => axios({
  method: 'post',
  url: '/api/v1/login',
  data,
  timeout: 4000,
});
