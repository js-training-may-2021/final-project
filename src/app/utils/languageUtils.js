import { useTranslation } from 'react-i18next';

export const setDefaultLanguage = () => {
  const defaultLanguage = 'ru';

  if (localStorage.getItem('language') === null) {
    localStorage.setItem('language', defaultLanguage);
  }
};

export const getCurrentLanguage = () => localStorage.getItem('language');

export const toggleLanguage = () => {
  const currentLanguage = localStorage.getItem('language');
  const newLanguage = currentLanguage === 'ru' ? 'en' : 'ru';
  localStorage.setItem('language', newLanguage);

  return newLanguage;
};

export const loc = (text) => {
  const { t } = useTranslation();
  return t(text);
};
