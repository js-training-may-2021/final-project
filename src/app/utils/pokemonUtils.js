import _ from 'lodash';

export const isFound = ({ name, id }, searchText) => {
  const searchBy = (attr, text) => String(attr).includes(text);
  const normalizedText = _.trimStart(searchText.trim().toLowerCase(), '#');
  return searchBy(name, normalizedText) || searchBy(id, normalizedText);
};

export const getNextId = (currentId) => {
  const nextId = currentId + 1 === 803 ? 10001 : currentId + 1;
  return nextId === 10148 ? 1 : nextId;
};

export const getPreviousId = (currentId) => {
  const prevId = currentId - 1 === 0 ? 10147 : currentId - 1;
  return prevId === 10000 ? 802 : prevId;
};

export const mappingSortPokemons = {
  idAsc: (array) => _.orderBy(array, 'id', 'asc'),
  idDesc: (array) => _.orderBy(array, 'id', 'desc'),
  nameAsc: (array) => _.orderBy(array, 'name', 'asc'),
  nameDesc: (array) => _.orderBy(array, 'name', 'desc'),
};
