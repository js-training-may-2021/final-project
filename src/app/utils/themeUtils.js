import { useSelector } from 'react-redux';

export const setDefaultTheme = () => {
  const defaultTheme = 'light';

  if (localStorage.getItem('theme') === null) {
    localStorage.setItem('theme', defaultTheme);
  }

  return localStorage.getItem('theme');
};

export const toggleTheme = () => {
  const currentTheme = localStorage.getItem('theme');
  const newTheme = currentTheme === 'light' ? 'dark' : 'light';
  localStorage.setItem('theme', newTheme);

  return newTheme;
};

export const isDarkTheme = () => {
  const currentTheme = useSelector((state) => state.uiState.theme);
  return currentTheme === 'dark';
};
