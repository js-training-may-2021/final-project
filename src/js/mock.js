export const pokemons = [
  {
    name: 'bulbasaur',
    id: 1,
    isCaught: false,
    catchDate: null,
  },
  {
    name: 'ivysaur',
    id: 2,
    isCaught: true,
    catchDate: '23 June 2021',
  },
  {
    name: 'venusaur',
    id: 3,
    isCaught: true,
    catchDate: '23 June 2021'
  },
  {
    name: 'charmander',
    id: 4,
    isCaught: false,
    catchDate: null,
  },
  {
    name: 'charmeleon',
    id: 5,
    isCaught: true,
    catchDate: '23 June 2021'
  },
  {
    name: 'charizard',
    id: 6,
    isCaught: true,
    catchDate: '23 June 2021'
  },
  {
    name: 'squirtle',
    id: 7,
    isCaught: false,
    catchDate: null,
  },
  {
    name: 'wartortle',
    id: 8,
    isCaught: true,
    catchDate: '23 June 2021'
  },
  {
    name: 'blastoise',
    id: 9,
    isCaught: true,
    catchDate: '23 June 2021'
  },
  {
    name: 'caterpie',
    id: 10,
    isCaught: false,
    catchDate: null,
  },
  {
    name: 'metapod',
    id: 11,
    isCaught: false,
    catchDate: null,
  }];  

export const uncaughtPokemon = pokemons[0];
export const caughtPokemon = pokemons[1];
