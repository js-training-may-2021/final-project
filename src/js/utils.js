import {createMemoryHistory} from 'history';
import configureStore from 'redux-mock-store';

export const POKEMONS_PER_PAGE = 10;
const START_PAGE = 1;

let currentPage = START_PAGE;

export const getCurrentPage = () => {
	return currentPage++;
};

export const AppRoute = {
  MAIN: '/',
  DETAIL: '/detail/:id',
  CAUGHT: '/caught'
};

const monthNames = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];

export const getCatchDate = () => {
  const date = new Date();

  const day = date.getDate();
  const month = monthNames[date.getMonth()];
  const year = date.getFullYear();

  return `${day} ${month} ${year}`
};

export const noop = () => {};
export const history = createMemoryHistory();
const mockStore = configureStore();
export const testStore = mockStore({});
