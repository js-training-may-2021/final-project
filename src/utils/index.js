export const firstLetterToUpperCase = (str) => {
  if (str === undefined) {
    return str;
  } else {
    return str
      .toLowerCase()
      .split(" ")
      .map((s) => s.charAt(0).toUpperCase() + s.substring(1))
      .join(" ");
  }
};

export const formatDate = (dateInput) => {
  return dateInput.toLocaleString("ru-RU");
};

export const padToThree = (number) => number <= 999 ? `00${number}`.slice(-3) : number;

export const handleErrors = (response) => {
  if (!response.ok) {
    throw Error(`${response.status}: ${response.statusText}`);
  }
  return response;
};
