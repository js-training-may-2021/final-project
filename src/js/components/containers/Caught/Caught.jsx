import React, {Fragment, useEffect} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {getCaughtPokemons} from '../../../store/data/selectors.js';
import {Operation} from '../../../store/data/data.js';
import Header from '../../presentationals/Header/Header.jsx';
import Footer from '../../presentationals/Footer/Footer.jsx';
import List from '../../presentationals/List/List.jsx';
import ErrorMessage from '../../presentationals/ErrorMessage/ErrorMessage.jsx';
import Loader from '../../presentationals/Loader/Loader.jsx';

const Caught = (props) => {
  const {
    caughtPokemons,
		isLoading,
		errorMessage,
    onCardClick,
    onButtonClick,
		loadPokemons
  } = props;

	useEffect(() => {
		loadPokemons();
	}, []);

	return (
		<Fragment>
			<Header />
			<main className="page__main">
				<h1 className="page__title">CAUGHT POKEMONS</h1>
				{!isLoading && !errorMessage && caughtPokemons.length === 0 ?
					<p className="page__text">You haven&apos;t caught any pokemons yet.</p> :
					<List
						pokemons={caughtPokemons}
						onCardClick={onCardClick}
						onButtonClick={onButtonClick}
					/>}
        {isLoading && <Loader />}
        {errorMessage && 
          <ErrorMessage
            message={errorMessage}
          />}
			</main>
			<Footer />
		</Fragment>
	);
};

Caught.propTypes = {
	caughtPokemons: PropTypes.arrayOf(PropTypes.shape({
		id: PropTypes.number.isRequired,
		name: PropTypes.string.isRequired,
		isCaught: PropTypes.bool.isRequired,
		catchDate: PropTypes.string,
	})),
	isLoading: PropTypes.bool.isRequired,
	errorMessage: PropTypes.string,
  onCardClick: PropTypes.func.isRequired,
  onButtonClick: PropTypes.func.isRequired,
  loadPokemons: PropTypes.func,
};

const mapStateToProps = state => ({
	caughtPokemons: getCaughtPokemons(state),
});

const mapDispatchToProps = dispatch => ({
	loadPokemons() {
    dispatch(Operation.loadCaughtPokemons());
	}
});

export default connect(mapStateToProps, mapDispatchToProps)(Caught);
