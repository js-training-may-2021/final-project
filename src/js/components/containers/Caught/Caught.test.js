import React from 'react';
import renderer from 'react-test-renderer';
import {Router} from 'react-router-dom';
import {Provider} from 'react-redux';
import {Caught} from './Caught.jsx';
import {history, noop, testStore} from '../../../utils.js';
import {pokemons} from '../../../mock.js';

it(`Render Caught page with pokemons`, () => {
  const tree = renderer
    .create(
      <Provider store={testStore}>
        <Router history={history}>
          <Caught
            caughtPokemons={pokemons}
            isLoading={false}
            errorMessage={null}
            onCardClick={noop}
            onButtonClick={noop}
            loadPokemons={noop}
          />
          </Router>
        </Provider>)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it(`Render Caught page with Loader`, () => {
  const tree = renderer
    .create(
      <Provider store={testStore}>
        <Router history={history}>
          <Caught
            caughtPokemons={[]}
            isLoading={true}
            errorMessage={null}
            onCardClick={noop}
            onButtonClick={noop}
            loadPokemons={noop}
          />
          </Router>
        </Provider>)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it(`Render Caught page with ErrorMessage`, () => {
  const tree = renderer
    .create(
      <Provider store={testStore}>
        <Router history={history}>
          <Caught
            caughtPokemons={[]}
            isLoading={false}
            errorMessage={'Failed to load pokemons. Please try again later.'}
            onCardClick={noop}
            onButtonClick={noop}
            loadPokemons={noop}
          />
          </Router>
        </Provider>)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
