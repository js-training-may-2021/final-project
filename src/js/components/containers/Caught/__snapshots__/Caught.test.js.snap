// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Caught page with ErrorMessage 1`] = `
Array [
  <header
    className="page__header header"
  >
    <nav
      className="header__nav-wrap"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          POKEDEX
        </a>
      </div>
      <ul
        className="header__menu menu"
      >
        <li
          className="menu__item"
        >
          <a
            aria-current="page"
            className="menu__link menu__link--selected"
            href="/"
            onClick={[Function]}
            style={Object {}}
          >
            ALL
          </a>
        </li>
        <li
          className="menu__item"
        >
          <a
            aria-current={null}
            className="menu__link"
            href="/caught"
            onClick={[Function]}
          >
            CAUGHT
          </a>
        </li>
      </ul>
    </nav>
  </header>,
  <main
    className="page__main"
  >
    <h1
      className="page__title"
    >
      CAUGHT POKEMONS
    </h1>
    <ul
      className="page__list list"
    />
    <p
      className="page__text page__text--message"
    >
      Failed to load pokemons. Please try again later.
    </p>
  </main>,
  <footer
    className="page__footer footer"
  >
    <div
      className="footer__nav-wrap"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          POKEDEX
        </a>
      </div>
      <ul
        className="header__menu menu"
      >
        <li
          className="menu__item"
        >
          <a
            aria-current="page"
            className="menu__link menu__link--selected"
            href="/"
            onClick={[Function]}
            style={Object {}}
          >
            ALL
          </a>
        </li>
        <li
          className="menu__item"
        >
          <a
            aria-current={null}
            className="menu__link"
            href="/caught"
            onClick={[Function]}
          >
            CAUGHT
          </a>
        </li>
      </ul>
    </div>
  </footer>,
]
`;

exports[`Render Caught page with Loader 1`] = `
Array [
  <header
    className="page__header header"
  >
    <nav
      className="header__nav-wrap"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          POKEDEX
        </a>
      </div>
      <ul
        className="header__menu menu"
      >
        <li
          className="menu__item"
        >
          <a
            aria-current="page"
            className="menu__link menu__link--selected"
            href="/"
            onClick={[Function]}
            style={Object {}}
          >
            ALL
          </a>
        </li>
        <li
          className="menu__item"
        >
          <a
            aria-current={null}
            className="menu__link"
            href="/caught"
            onClick={[Function]}
          >
            CAUGHT
          </a>
        </li>
      </ul>
    </nav>
  </header>,
  <main
    className="page__main"
  >
    <h1
      className="page__title"
    >
      CAUGHT POKEMONS
    </h1>
    <ul
      className="page__list list"
    />
    <p
      className="page__text page__text--message"
    >
      Loading...
    </p>
  </main>,
  <footer
    className="page__footer footer"
  >
    <div
      className="footer__nav-wrap"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          POKEDEX
        </a>
      </div>
      <ul
        className="header__menu menu"
      >
        <li
          className="menu__item"
        >
          <a
            aria-current="page"
            className="menu__link menu__link--selected"
            href="/"
            onClick={[Function]}
            style={Object {}}
          >
            ALL
          </a>
        </li>
        <li
          className="menu__item"
        >
          <a
            aria-current={null}
            className="menu__link"
            href="/caught"
            onClick={[Function]}
          >
            CAUGHT
          </a>
        </li>
      </ul>
    </div>
  </footer>,
]
`;

exports[`Render Caught page with pokemons 1`] = `
Array [
  <header
    className="page__header header"
  >
    <nav
      className="header__nav-wrap"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          POKEDEX
        </a>
      </div>
      <ul
        className="header__menu menu"
      >
        <li
          className="menu__item"
        >
          <a
            aria-current="page"
            className="menu__link menu__link--selected"
            href="/"
            onClick={[Function]}
            style={Object {}}
          >
            ALL
          </a>
        </li>
        <li
          className="menu__item"
        >
          <a
            aria-current={null}
            className="menu__link"
            href="/caught"
            onClick={[Function]}
          >
            CAUGHT
          </a>
        </li>
      </ul>
    </nav>
  </header>,
  <main
    className="page__main"
  >
    <h1
      className="page__title"
    >
      CAUGHT POKEMONS
    </h1>
    <ul
      className="page__list list"
    >
      <li
        className="list__card card"
        onClick={[Function]}
      >
        <a
          className="card__link-wrap"
          href="/detail/1"
          onClick={[Function]}
        >
          <img
            alt="Bulbasaur"
            className="card__img"
            src="test-file-stub"
          />
          <h2
            className="card__title"
          >
            Bulbasaur
          </h2>
        </a>
        <button
          className="card__button button"
          onClick={[Function]}
          type="button"
        >
          Catch!
        </button>
      </li>
      <li
        className="list__card card card--caught"
        onClick={[Function]}
      >
        <a
          className="card__link-wrap"
          href="/detail/2"
          onClick={[Function]}
        >
          <img
            alt="Ivysaur"
            className="card__img"
            src="test-file-stub"
          />
          <h2
            className="card__title"
          >
            Ivysaur
          </h2>
        </a>
        <button
          className="card__button button button--checked"
          onClick={[Function]}
          type="button"
        >
          Release
        </button>
      </li>
      <li
        className="list__card card card--caught"
        onClick={[Function]}
      >
        <a
          className="card__link-wrap"
          href="/detail/3"
          onClick={[Function]}
        >
          <img
            alt="Venusaur"
            className="card__img"
            src="test-file-stub"
          />
          <h2
            className="card__title"
          >
            Venusaur
          </h2>
        </a>
        <button
          className="card__button button button--checked"
          onClick={[Function]}
          type="button"
        >
          Release
        </button>
      </li>
      <li
        className="list__card card"
        onClick={[Function]}
      >
        <a
          className="card__link-wrap"
          href="/detail/4"
          onClick={[Function]}
        >
          <img
            alt="Charmander"
            className="card__img"
            src="test-file-stub"
          />
          <h2
            className="card__title"
          >
            Charmander
          </h2>
        </a>
        <button
          className="card__button button"
          onClick={[Function]}
          type="button"
        >
          Catch!
        </button>
      </li>
      <li
        className="list__card card card--caught"
        onClick={[Function]}
      >
        <a
          className="card__link-wrap"
          href="/detail/5"
          onClick={[Function]}
        >
          <img
            alt="Charmeleon"
            className="card__img"
            src="test-file-stub"
          />
          <h2
            className="card__title"
          >
            Charmeleon
          </h2>
        </a>
        <button
          className="card__button button button--checked"
          onClick={[Function]}
          type="button"
        >
          Release
        </button>
      </li>
      <li
        className="list__card card card--caught"
        onClick={[Function]}
      >
        <a
          className="card__link-wrap"
          href="/detail/6"
          onClick={[Function]}
        >
          <img
            alt="Charizard"
            className="card__img"
            src="test-file-stub"
          />
          <h2
            className="card__title"
          >
            Charizard
          </h2>
        </a>
        <button
          className="card__button button button--checked"
          onClick={[Function]}
          type="button"
        >
          Release
        </button>
      </li>
      <li
        className="list__card card"
        onClick={[Function]}
      >
        <a
          className="card__link-wrap"
          href="/detail/7"
          onClick={[Function]}
        >
          <img
            alt="Squirtle"
            className="card__img"
            src="test-file-stub"
          />
          <h2
            className="card__title"
          >
            Squirtle
          </h2>
        </a>
        <button
          className="card__button button"
          onClick={[Function]}
          type="button"
        >
          Catch!
        </button>
      </li>
      <li
        className="list__card card card--caught"
        onClick={[Function]}
      >
        <a
          className="card__link-wrap"
          href="/detail/8"
          onClick={[Function]}
        >
          <img
            alt="Wartortle"
            className="card__img"
            src="test-file-stub"
          />
          <h2
            className="card__title"
          >
            Wartortle
          </h2>
        </a>
        <button
          className="card__button button button--checked"
          onClick={[Function]}
          type="button"
        >
          Release
        </button>
      </li>
      <li
        className="list__card card card--caught"
        onClick={[Function]}
      >
        <a
          className="card__link-wrap"
          href="/detail/9"
          onClick={[Function]}
        >
          <img
            alt="Blastoise"
            className="card__img"
            src="test-file-stub"
          />
          <h2
            className="card__title"
          >
            Blastoise
          </h2>
        </a>
        <button
          className="card__button button button--checked"
          onClick={[Function]}
          type="button"
        >
          Release
        </button>
      </li>
      <li
        className="list__card card"
        onClick={[Function]}
      >
        <a
          className="card__link-wrap"
          href="/detail/10"
          onClick={[Function]}
        >
          <img
            alt="Caterpie"
            className="card__img"
            src="test-file-stub"
          />
          <h2
            className="card__title"
          >
            Caterpie
          </h2>
        </a>
        <button
          className="card__button button"
          onClick={[Function]}
          type="button"
        >
          Catch!
        </button>
      </li>
      <li
        className="list__card card"
        onClick={[Function]}
      >
        <a
          className="card__link-wrap"
          href="/detail/11"
          onClick={[Function]}
        >
          <img
            alt="Metapod"
            className="card__img"
            src="test-file-stub"
          />
          <h2
            className="card__title"
          >
            Metapod
          </h2>
        </a>
        <button
          className="card__button button"
          onClick={[Function]}
          type="button"
        >
          Catch!
        </button>
      </li>
    </ul>
  </main>,
  <footer
    className="page__footer footer"
  >
    <div
      className="footer__nav-wrap"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          POKEDEX
        </a>
      </div>
      <ul
        className="header__menu menu"
      >
        <li
          className="menu__item"
        >
          <a
            aria-current="page"
            className="menu__link menu__link--selected"
            href="/"
            onClick={[Function]}
            style={Object {}}
          >
            ALL
          </a>
        </li>
        <li
          className="menu__item"
        >
          <a
            aria-current={null}
            className="menu__link"
            href="/caught"
            onClick={[Function]}
          >
            CAUGHT
          </a>
        </li>
      </ul>
    </div>
  </footer>,
]
`;
